design/specs/specs-null-values.md

```diff
@@ -0,0 +1,128 @@
+
+#### Scenario 1a: The user identifies a NULL value in a table
+
+##### Steps for 1a
+
+1. The user opens a table that contains NULL values
+2. The user identifies a cell that contains a NULL value
+    - The cell contains the word NULL in all uppercase letters
+    - The cell has a lighter background-color
+    - The cell has a muted text color
+    - The cell has a different font style (italic)
+
+#### Scenario 1b: The user identifies a NULL value in a record preview
+
+##### Steps for 1b
+
+1. The user edits a value in a column that links to records from another table
+2. The user sees the list of records displayed in the record selection dropdown
+3. The user identifies a record that contains a NULL value
+    - The field contains the word NULL in all uppercase letters
+    - The text has a muted text color
+    - The text has a different font style (italic)
+
+### Scenario 2: User edits a NULL Value
+
+#### Scenario 2a: The data type is text
+
+##### Steps for 2a
+
+1. The user edits a NULL value in a column that is set to text data type
+2. The user double-clicks the value cell to enter edit mode
+3. The user enters a value and leaves edit mode by pressing the enter key or clicking outside of the active cell
+4. The NULL value is replaced by the new value
+
+#### Scenario 2b: The data type is text, and the value is cleared
+
+#### Steps for 2b
+
+1. The user edits a NULL value in a column that is set to text data type
+2. The user double-clicks the value cell to enter edit mode
+3. In edit-mode, a text input box is displayed with a placeholder set to 'NULL'
+4. The user types in a value and the placeholder text is replaced by the new typed-in value
+5. The user clears the new value by hitting the backspace key until the value is completely cleared
+6. The user exits edit-mode
+7. The NULL value is replaced by an empty string and the cell is blank
+
+#### Scenario 2c: The data type is boolean with checkbox display enabled
+
+##### Steps for 2c
+
+1. The user edits a NULL value in a column that is set to boolean type with the checkbox display property enabled
+    - The field contains a checkbox in an indeterminate state
+    - Hovering over the NULL checkbox will display a tooltip with the 'NULL' legend.
+2. The user clicks on the checkbox, going from indeterminate (NULL) to checked (TRUE).
+3. The user can click on the checkbox again to mark it as false and go back to NULL.
+
+#### Scenario 2d: The data type is number
+
+##### Steps for 2d
+
+1. The user edits a NULL value in a column set to number type.
+2. The user double-clicks the cell to enter edit mode.
+3. The user exists in edit mode without making any changes.
+4. The NULL value remains.
+
+### Scenario 3: User sets a value to NULL
+
+#### Scenario 3a: The column accepts NULL values
+
+##### Steps for 3a
+
+1. The user selects a cell in a column that contains a value.
+2. The user clears the content of the selected cell.
+    - By pressing the delete key.
```

Comment:
I'm a little confused about how this will work with our current "spreadsheet like" editing for tables. Right now, the user only needs to click on a cell to edit it. This seems like it's introducing a pattern where the user needs to double click on a cell to edit it. I think this would hurt rapid data entry. I'd rather remove the feature where pressing `Delete`/`Backspace` would clear the cell and keep the single-click-to-edit pattern.

Reply:
> Right now, the user only needs to click on a cell to edit it
With the current master branch frontend interface, I have to double click to enter edit mode, clicking it once enters selection mode. Is this something part of the [Spreadsheet like editing milestone](https://github.com/centerofci/mathesar/milestone/100)

 Though we use a `column based` database and not a `row based` spreadsheet, we would still be needing a selection mode to reference, copy or drag and drop it as a reference. For any operation on a row rather than a row value, we would need a selection mode.

Reply:
I agree, I'm not arguing against the selection indicator. I don't see why clicking on a cell once can't do both actions - it can select the cell and also process any keystrokes while selected as editing the contents of the cell.

Reply:
Makes sense, I would love to have something like this tbh, but I think the input element will mess up with the functionalities of the selection mode. For example, drag and drop would end up moving the text rather than the element itself.

Reply:
I'm not sure if that's necessarily the case. When I've implemented drag and drop in frontend, I think I had the flexibility to bubble the drag target up to the parent element. @seancolsen or @pavish would know better though.

Reply:
However we spec this, I'm confident we'll be able to implement it. 

@silentninja is correct that the master branch currently requires double clicking before I can edit a cell. But I agree with @kgodey that we should change that behavior to be more consistent with spreadsheets. On a spreadsheet I can begin entering text even when the cell is not in edit mode. As soon as the first keystroke begins the cell switches to edit mode and begins replacing the contents with the key I just entered. That should be easy. I'm not seeing the relevant of drag and drop for this workflow, but I'm sure there will be edge cases to consider as we implement this. The good thing is that "spreadsheet like" editing is already fairly well-defined by other products so we just need to follow that UX. The fact that Google has come so far with their Sheets product on the web is part of what gives me confidence that we'll be able to implement our specs.

Reply:
I totally missed the fact that the cell turns into edit mode on keypress, I was thinking of not having an edit state visual cue at all. If that's the case then I think we don't have to remove the cell deletion upon the `Delete` keypress feature.

So an example scenario would be:
1. User highlights the cell with a single click, pressing the `delete` or `backspace` key will delete the cell state and convert it to a null value. 
2. Then the user presses any key other than `delete` or `backspace` the cell enters into an edit mode.  Pressing `delete` or `backspace` will delete the text inside the cell.

I think the same can be applied for boolean types too. The `checkbox` should be smaller than the cell
1. Clicking outside the `checkbox` would select the cell and pressing `delete` would make the cell null
2. Clicking inside the `checkbox` would toggle the value of the checkbox, similar to spreadsheets.

The only issue is with `dropdown's` though, we could have a null value as a dropdown option.

Reply:
What @mathemancer described here would be the correct behavior.
> As soon as the first keystroke begins the cell switches to edit mode and begins replacing the contents with the key I just entered. 

I think @silentninja is describing the same. I will update the spect to reflect this.

Reply:
Sorry about the misplaced comment earlier. Didn't read down all the way. So... regarding: 
> User highlights the cell with a single click, pressing the `delete` or `backspace` key will delete the cell state and convert it to a null value.
> Then the user presses any key other than `delete` or `backspace` the cell enters into an edit mode. Pressing `delete` or `backspace` will delete the text inside the cell.

Maybe too late and this is probably the right way to do it, but just intuitively, the idea of having to type something and then delete it to get an empty cell just feels "hacky" in sort of a bad way. Maybe there is a setting, but I'd rather, for example, do `⌘delete` or `⌃delete` for NULL and normal `delete` for empty string, for example...

Reply:
@Joi Upon single click, the cell gets selected, sort of like command mode in text editors. 

When any key excluding cell shortcut keys like delete, backspace is pressed, the cell goes into `edit mode`, which changes the appearance of cell from disabled grey to a text box, so keystrokes work like a shortcut for `enter key`. 

`delete` key will delete the words when in `edit mode`,  when in `command` mode it will delete the cell state.

Reply:
@silentninja - ah, I see. So if I hit `delete` in `edit mode`, I can end up with an empty string without having to type anything other than `delete`? That makes sense. I misunderstood the interaction.

Reply:
@Joi 

> So if I hit `delete` in `edit mode`, I can end up with an empty string without having to type anything other than `delete`?

If I'm understanding you correctly, I don't think this spec is headed (or should head) in that direction.

**When you're in edit mode**, you're editing text. You have a cursor. You can type new words. You can use the arrow keys to move your cursor. You can delete what you typed with keys like `Delete` and `Backspace`. Keys like `Delete` and `Backspace` (as well as key combinations with Command, Control, Shift, Option, Alt) have well established behavior (albeit with subtle differences across platforms) for editing text, and I don't think we should mess with that behavior. If you want to change the value of a cell from "Foo" to "" (empty string), then you'll need to double click to edit the cell. At this point your cursor will be placed _somewhere_ in the cell. We haven't defined that position yet. (For example LibreOffice places your cursor exactly where you clicked while Google Sheets places it at the end of the text.) In any case, at this point you'll need to manually _delete all the text_. There are may ways to do that. For example on Mac OS, `command+delete` will delete all the text between the cursor and the start of the line. That would delete all the text, but only if the cursor is at the end. Or `command+a` and then `delete` would select all and delete. Finally you'll need to press `Enter` to save. 

**When you're in select mode** you can use the arrow keys to move your selection between cells. You can press `Delete` or `Backspace` to set the cell value to NULL. There is no way to set the cell value to an empty string without entering edit mode and manually deleting all the text.

In general we're trying to discourage the use of empty string and encourage the use of NULL. 

Does my description make sense? And do you have any concerns with the direction we're heading? If so, it might be helpful for us to better understand the use case for empty string instead of NULL.

Reply:
Further clarification...

When editing text:

- On Mac OS, pressing `Delete` deletes one character to the left. Pressing `Fn-Delete` deletes one character to the right.
- On Windows and Linux, pressing `Backspace` deletes one character to the left. Pressing `Delete` deletes one character to the right.

So when @silentninja said _"`delete` key will delete the words when in `edit mode`"_, I interpret that to mean _"the `delete` key will have its normal behavior (described above) when in `edit mode`"_.

Reply:
@Joi Sorry for the confusion I meant to say that delete will have its normal behaviour in `edit mode` which is to delete the letter not the word @seancolsen explains it much better.

Reply:
@seancolsen's definition gives a clear idea of change in behaviour between select mode and edit mode.

There is a scenario we need to take into account regarding rapid data entry (@kgodey had a similar concern):

When the user is already in `edit` mode on a particular cell and then enters 'Tab', the active cell shifts and the immediate next cell enters into `edit` mode directly. This behaviour is similar to spreadsheets and it is what we have currently. When users are entering data rapidly, this would be their normal workflow.

In this scenario, users would never have to enter the `select` mode. So, setting NULL on a cell would be tiring. They would have to esc into `select` mode and set NULL and re-enter `edit` mode on the next cell.

@ghislaineguerin We probably need a keyboard combination to set NULL while the cell is in `edit` mode. Other spreadsheet applications do not have this problem (since they don't usually have the concept of NULL), so this would be something we can research and decide for ourselves.

Reply:
@pavish

## Product behavior

- I'm not sure we need a keyboard shortcut to set the cell value to NULL while in edit mode.
- In a spreadsheet, when you press `Tab` your selection moves to the right. But after that, you end up in select mode not edit mode. So if you want to set several adjacent cells in one column to NULL, you'd be able to do that by pressing `Delete`, `Tab`, `Delete`, `Tab` and so on.
- I can see how in some cases it would be useful to have a keyboard shortcut for setting a cell value to NULL while in edit mode. But it also introduces other UX questions that I don't think are worth spending the time to answer at this point. What happens _after_ I execute that keyboard shortcut? Am I still in edit mode? What if I mistakenly execute that keyboard shortcut? Will I be able to _undo_? It might be harder to support undo if we move to select mode. I suggest that we defer the _"set to NULL in edit mode"_ keyboard shortcut feature until a time when it becomes a more pressing need.

## Nomenclature

As an aside...

- I think in this conversation and spec (and others like it) it might be helpful if we make an effort refer to keys by the `key` value of the DOM `KeyboardEvent` (instead of by the text printed on the physical keyboard).
- On Mac OS, pressing the key labeled "delete" will send a `Backspace` event (and delete one character to the left). Pressing the keys labeled `Fn+delete` will send a `Delete` event and delete one character to the right.
- On Windows and Linux the physical key labels match their DOM event `key` values.
- That means that when we say _"press `delete`"_ in this discussion, it's somewhat ambiguous.
- If you execute `window.addEventListener('keydown', e => console.log(e.key))` in your browser console then you can press keys and see their respective DOM `key` values.

Reply:
@seancolsen that's a great suggestion on nomenclature, I've been struggling with maintaining consistency with that. I'll default to DOM `key` values from now on.

Reply:
The process @seancolsen described matches what I'd expect personally.  I also agree that it's not that necessary to have a keyboard shortcut while in edit mode to set `NULL`.